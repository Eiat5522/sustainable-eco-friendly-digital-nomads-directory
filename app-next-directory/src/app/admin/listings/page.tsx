'use client';

import { Badge } from '@/components/ui/badge';
import { Button } from '@/components/ui/Button';
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '@/components/ui/card';
import {
    DropdownMenu,
    DropdownMenuContent,
    DropdownMenuItem,
    DropdownMenuLabel,
    DropdownMenuSeparator,
    DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';
import { Input } from '@/components/ui/input';
import {
    CheckCircle,
    Clock,
    Eye,
    Flag,
    Home,
    MapPin,
    MoreHorizontal,
    Search,
    Star,
    XCircle
} from 'lucide-react';
import React, { useEffect, useState } from 'react';

interface ListingStats {
  totalListings: number;
  publishedListings: number;
  pendingListings: number;
  flaggedListings: number;
  averageRating: number;
  totalViews: number;
}

interface Listing {
  id: string;
  title: string;
  type: 'accommodation' | 'coworking' | 'restaurant' | 'activity';
  status: 'published' | 'pending' | 'flagged' | 'draft' | 'rejected';
  location: {
    city: string;
    country: string;
  };
  rating: number;
  reviewCount: number;
  views: number;
  createdAt: string;
  updatedAt: string;
  author: {
    id: string;
    name: string;
    email: string;
  };
  sustainabilityScore: number;
  features: string[];
  price?: {
    amount: number;
    currency: string;
    period: string;
  };
}

export default function ListingsPage() {
  const [listingStats, setListingStats] = useState<ListingStats>({
    totalListings: 0,
    publishedListings: 0,
    pendingListings: 0,
    flaggedListings: 0,
    averageRating: 0,
    totalViews: 0,
  });

  const [listings, setListings] = useState<Listing[]>([]);
  const [loading, setLoading] = useState(true);
  const [searchTerm, setSearchTerm] = useState('');
  const [typeFilter, setTypeFilter] = useState<'all' | 'accommodation' | 'coworking' | 'restaurant' | 'activity'>('all');
  const [statusFilter, setStatusFilter] = useState<'all' | 'published' | 'pending' | 'flagged' | 'draft' | 'rejected'>('all');

  useEffect(() => {
    fetchListings();
  }, []);

  const fetchListings = async () => {
    try {
      setLoading(true);

      // TODO: Replace with actual API calls
      const mockStats: ListingStats = {
        totalListings: 156,
        publishedListings: 142,
        pendingListings: 8,
        flaggedListings: 3,
        averageRating: 4.2,
        totalViews: 15420,
      };

      const mockListings: Listing[] = [
        {
          id: '1',
          title: 'Eco-Friendly Co-working Space in Bali',
          type: 'coworking',
          status: 'published',
          location: {
            city: 'Ubud',
            country: 'Indonesia',
          },
          rating: 4.8,
          reviewCount: 24,
          views: 342,
          createdAt: '2024-11-15T10:30:00Z',
          updatedAt: '2024-12-18T14:20:00Z',
          author: {
            id: 'user1',
            name: 'John Doe',
            email: 'john@example.com',
          },
          sustainabilityScore: 92,
          features: ['Solar Power', 'Recycling Program', 'Local Materials'],
          price: {
            amount: 25,
            currency: 'USD',
            period: 'day',
          },
        },
        {
          id: '2',
          title: 'Solar-Powered Accommodation in Costa Rica',
          type: 'accommodation',
          status: 'flagged',
          location: {
            city: 'San José',
            country: 'Costa Rica',
          },
          rating: 3.2,
          reviewCount: 8,
          views: 89,
          createdAt: '2024-12-10T15:45:00Z',
          updatedAt: '2024-12-19T09:30:00Z',
          author: {
            id: 'user2',
            name: 'Jane Smith',
            email: 'jane@example.com',
          },
          sustainabilityScore: 78,
          features: ['Solar Power', 'Rainwater Harvesting'],
          price: {
            amount: 85,
            currency: 'USD',
            period: 'night',
          },
        },
        {
          id: '3',
          title: 'Sustainable Farm-to-Table Restaurant',
          type: 'restaurant',
          status: 'pending',
          location: {
            city: 'Lisbon',
            country: 'Portugal',
          },
          rating: 4.5,
          reviewCount: 15,
          views: 123,
          createdAt: '2024-12-18T09:15:00Z',
          updatedAt: '2024-12-19T11:00:00Z',
          author: {
            id: 'user3',
            name: 'Alice Johnson',
            email: 'alice@example.com',
          },
          sustainabilityScore: 88,
          features: ['Organic Ingredients', 'Zero Waste', 'Local Sourcing'],
        },
      ];

      setListingStats(mockStats);
      setListings(mockListings);
    } catch (error) {
      console.error('Error fetching listings:', error);
    } finally {
      setLoading(false);
    }
  };

  const filteredListings = listings.filter(listing => {
    const matchesSearch = listing.title.toLowerCase().includes(searchTerm.toLowerCase()) ||
                         listing.location.city.toLowerCase().includes(searchTerm.toLowerCase()) ||
                         listing.location.country.toLowerCase().includes(searchTerm.toLowerCase());
    const matchesType = typeFilter === 'all' || listing.type === typeFilter;
    const matchesStatus = statusFilter === 'all' || listing.status === statusFilter;
    return matchesSearch && matchesType && matchesStatus;
  });
  const getTypeBadgeVariant = (type: string): 'default' | 'secondary' | 'outline' | 'destructive' => {
    switch (type) {
      case 'accommodation': return 'default';
      case 'coworking': return 'secondary';
      case 'restaurant': return 'outline';
      case 'activity': return 'secondary';
      default: return 'default';
    }
  };

  const getStatusBadgeVariant = (status: string): 'default' | 'secondary' | 'outline' | 'destructive' => {
    switch (status) {
      case 'published': return 'default';
      case 'pending': return 'secondary';
      case 'flagged': return 'destructive';
      case 'draft': return 'outline';
      case 'rejected': return 'destructive';
      default: return 'default';
    }
  };

  const getStatusIcon = (status: string) => {
    switch (status) {
      case 'published': return <CheckCircle className="h-4 w-4 text-green-500" />;
      case 'pending': return <Clock className="h-4 w-4 text-yellow-500" />;
      case 'flagged': return <Flag className="h-4 w-4 text-red-500" />;
      case 'rejected': return <XCircle className="h-4 w-4 text-red-500" />;
      default: return <Clock className="h-4 w-4 text-gray-500" />;
    }
  };

  const getTypeIcon = (type: string) => {
    switch (type) {
      case 'accommodation': return '🏠';
      case 'coworking': return '💼';
      case 'restaurant': return '🍽️';
      case 'activity': return '🎯';
      default: return '📋';
    }
  };

  const handleListingAction = async (listingId: string, action: 'publish' | 'flag' | 'reject' | 'delete' | 'edit') => {
    // TODO: Implement actual listing management actions
    console.log(`${action} listing ${listingId}`);
  };

  if (loading) {
    return (
      <div className="flex items-center justify-center min-h-screen">
        <div className="text-lg">Loading listings...</div>
      </div>
    );
  }

  return (
    <div className="p-6 max-w-7xl mx-auto space-y-6">
      <div className="flex justify-between items-center">
        <div>
          <h1 className="text-3xl font-bold text-gray-900">Listings Management</h1>
          <p className="text-gray-600 mt-2">Manage directory listings, reviews, and content quality</p>
        </div>
        <Button onClick={fetchListings} disabled={loading}>
          Refresh Data
        </Button>
      </div>

      {/* Listing Statistics */}
      <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-6 gap-4">
        <Card>
          <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
            <CardTitle className="text-sm font-medium">Total Listings</CardTitle>
            <Home className="h-4 w-4 text-muted-foreground" />
          </CardHeader>
          <CardContent>
            <div className="text-2xl font-bold">{listingStats.totalListings}</div>
            <p className="text-xs text-muted-foreground">All listings</p>
          </CardContent>
        </Card>

        <Card>
          <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
            <CardTitle className="text-sm font-medium">Published</CardTitle>
            <CheckCircle className="h-4 w-4 text-green-500" />
          </CardHeader>
          <CardContent>
            <div className="text-2xl font-bold text-green-600">{listingStats.publishedListings}</div>
            <p className="text-xs text-muted-foreground">Live listings</p>
          </CardContent>
        </Card>

        <Card>
          <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
            <CardTitle className="text-sm font-medium">Pending</CardTitle>
            <Clock className="h-4 w-4 text-yellow-500" />
          </CardHeader>
          <CardContent>
            <div className="text-2xl font-bold text-yellow-600">{listingStats.pendingListings}</div>
            <p className="text-xs text-muted-foreground">Awaiting review</p>
          </CardContent>
        </Card>

        <Card>
          <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
            <CardTitle className="text-sm font-medium">Flagged</CardTitle>
            <Flag className="h-4 w-4 text-red-500" />
          </CardHeader>
          <CardContent>
            <div className="text-2xl font-bold text-red-600">{listingStats.flaggedListings}</div>
            <p className="text-xs text-muted-foreground">Need attention</p>
          </CardContent>
        </Card>

        <Card>
          <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
            <CardTitle className="text-sm font-medium">Avg Rating</CardTitle>
            <Star className="h-4 w-4 text-yellow-500" />
          </CardHeader>
          <CardContent>
            <div className="text-2xl font-bold text-yellow-600">{listingStats.averageRating}</div>
            <p className="text-xs text-muted-foreground">Overall quality</p>
          </CardContent>
        </Card>

        <Card>
          <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
            <CardTitle className="text-sm font-medium">Total Views</CardTitle>
            <Eye className="h-4 w-4 text-blue-500" />
          </CardHeader>
          <CardContent>
            <div className="text-2xl font-bold text-blue-600">{listingStats.totalViews.toLocaleString()}</div>
            <p className="text-xs text-muted-foreground">All time</p>
          </CardContent>
        </Card>
      </div>

      {/* Listings Management */}
      <Card>
        <CardHeader>
          <CardTitle>Directory Listings</CardTitle>
          <CardDescription>
            Manage all directory listings, their status, and content quality
          </CardDescription>
        </CardHeader>
        <CardContent>
          {/* Search and Filter Controls */}
          <div className="flex gap-4 mb-6">
            <div className="flex-1">
              <div className="relative">
                <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 h-4 w-4" />
                <Input
                  placeholder="Search listings..."
                  value={searchTerm}
                  onChange={(e: any) => setSearchTerm((e.currentTarget as HTMLInputElement).value)}
                  className="pl-10"
                />
              </div>
            </div>
            <select
              aria-label="Filter by listing type"
              value={typeFilter}
              onChange={(e: any) => setTypeFilter((e.currentTarget as HTMLSelectElement).value as any)}
              className="px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-green-500"
            >
              <option value="all">All Types</option>
              <option value="accommodation">Accommodation</option>
              <option value="coworking">Co-working</option>
              <option value="restaurant">Restaurant</option>
              <option value="activity">Activity</option>
            </select>
            <select
              aria-label="Filter by listing status"
              value={statusFilter}
              onChange={(e: any) => setStatusFilter((e.currentTarget as HTMLSelectElement).value as any)}
              className="px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-green-500"
            >
              <option value="all">All Status</option>
              <option value="published">Published</option>
              <option value="pending">Pending</option>
              <option value="flagged">Flagged</option>
              <option value="draft">Draft</option>
              <option value="rejected">Rejected</option>
            </select>
          </div>

          {/* Listings List */}
          <div className="space-y-4">
            {filteredListings.map((listing) => (
              <div key={listing.id} className="border rounded-lg p-4 hover:shadow-md transition-shadow">
                <div className="flex items-center justify-between">
                  <div className="flex items-center space-x-4">
                    <span className="text-2xl">{getTypeIcon(listing.type)}</span>
                    <div className="flex-1">
                      <div className="flex items-center gap-2">
                        <h3 className="font-semibold text-lg">{listing.title}</h3>
                        {getStatusIcon(listing.status)}
                      </div>
                      <p className="text-sm text-gray-600 flex items-center gap-2">
                        <MapPin className="h-3 w-3" />
                        {listing.location.city}, {listing.location.country}
                      </p>
                      <p className="text-xs text-gray-500">
                        by {listing.author.name} • Created: {new Date(listing.createdAt).toLocaleDateString()}
                      </p>
                      <div className="flex gap-2 mt-2">
                        {listing.features.slice(0, 3).map((feature, index) => (
                          <span key={index} className="text-xs bg-green-100 text-green-800 px-2 py-1 rounded">
                            {feature}
                          </span>
                        ))}
                      </div>
                    </div>
                  </div>
                  <div className="flex items-center space-x-6">
                    <div className="text-right">
                      <div className="flex items-center gap-2 text-sm text-gray-600">
                        <Star className="h-3 w-3 text-yellow-500" />
                        {listing.rating} ({listing.reviewCount} reviews)
                      </div>
                      <div className="text-sm text-gray-600">
                        {listing.views} views
                      </div>
                      <div className="text-sm font-medium text-green-600">
                        {listing.sustainabilityScore}% sustainable
                      </div>
                      {listing.price && (
                        <div className="text-sm text-gray-800 font-medium">
                          {listing.price.currency} {listing.price.amount}/{listing.price.period}
                        </div>
                      )}
                    </div>
                    <div className="flex gap-2">
                      <Badge variant={getTypeBadgeVariant(listing.type)}>
                        {listing.type}
                      </Badge>
                      <Badge variant={getStatusBadgeVariant(listing.status)}>
                        {listing.status}
                      </Badge>
                    </div>
                    <DropdownMenu>
                      <DropdownMenuTrigger asChild>
                        <Button variant="ghost" className="h-8 w-8 p-0">
                          <MoreHorizontal className="h-4 w-4" />
                        </Button>
                      </DropdownMenuTrigger>
                      <DropdownMenuContent align="end">
                        <DropdownMenuLabel>Actions</DropdownMenuLabel>
                        <DropdownMenuItem onClick={() => handleListingAction(listing.id, 'edit')}>
                          View/Edit Listing
                        </DropdownMenuItem>
                        <DropdownMenuSeparator />
                        {listing.status === 'pending' && (
                          <DropdownMenuItem
                            onClick={() => handleListingAction(listing.id, 'publish')}
                            className="text-green-600"
                          >
                            Approve & Publish
                          </DropdownMenuItem>
                        )}
                        {listing.status === 'published' && (
                          <DropdownMenuItem
                            onClick={() => handleListingAction(listing.id, 'flag')}
                            className="text-orange-600"
                          >
                            Flag for Review
                          </DropdownMenuItem>
                        )}
                        <DropdownMenuItem
                          onClick={() => handleListingAction(listing.id, 'reject')}
                          className="text-red-600"
                        >
                          Reject Listing
                        </DropdownMenuItem>
                        <DropdownMenuItem
                          onClick={() => handleListingAction(listing.id, 'delete')}
                          className="text-red-600"
                        >
                          Delete Listing
                        </DropdownMenuItem>
                      </DropdownMenuContent>
                    </DropdownMenu>
                  </div>
                </div>
              </div>
            ))}
          </div>

          {filteredListings.length === 0 && (
            <div className="text-center py-8 text-gray-500">
              No listings found matching your criteria.
            </div>
          )}
        </CardContent>
      </Card>
    </div>
  );
}
