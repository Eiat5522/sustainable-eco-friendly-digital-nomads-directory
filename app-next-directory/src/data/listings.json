[
  {
    "id": "sample-id",
    "name": "Sample Listing",
    "city": "Sample City",
    "category": "coworking",
    "address_string": "123 Sample Street",
    "description_short": "A short description.",
    "description_long": "A longer description of the sample listing.",
    "eco_focus_tags": ["eco-friendly", "sustainable"],
    "eco_notes_detailed": "Detailed notes about eco features.",
    "source_urls": ["https://example.com"],
    "primary_image_url": "https://example.com/image.jpg",
    "gallery_image_urls": ["https://example.com/gallery1.jpg"],
    "digital_nomad_features": ["fast wifi", "community events"],
    "last_verified_date": "2025-01-01"
  }
]
