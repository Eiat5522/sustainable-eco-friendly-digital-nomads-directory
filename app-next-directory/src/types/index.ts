export interface EcoTag {
  id: string;
  label: string;
  impact: 'high' | 'medium' | 'low';
}

export interface SearchFilters {
  category?: string[];
  ecoTags?: string[];
  minPrice?: number;
  maxPrice?: number;
  query?: string;
}

export interface FilterOption {
  id: string;
  label: string;
  count?: number;
  icon?: string;
  ecoImpact?: 'high' | 'medium' | 'low';
}

export interface FilterGroup {
  id: string;
  label: string;
  options: FilterOption[];
  multiSelect?: boolean;
  icon?: string;
}

// City type
export interface City {
  id: string;
  name: string;
  country: string;
  slug: string;
  description: string;
  description_short?: string;
  images?: string[];
  coordinates?: {
    latitude: number;
    longitude: number;
  };
  airQuality?: string;
  internetSpeed?: number;
  costOfLiving?: string;
  climate?: string;
  safety?: string;
  walkability?: string;
  sustainabilityInitiatives?: string[];
  nomadFeatures?: string[];
}

// Listing type
export interface Listing {
  id: string;
  name: string;
  slug: string;
  cityId: string;
  category: string;
  description_short: string;
  description_long: string;
  imageUrl: string;
  images?: string[];
  coordinates?: {
    latitude: number;
    longitude: number;
  };
  eco_features?: string[];
  amenities?: string[];
  price?: number;
  currency?: string;
  rating?: number;
  reviewCount?: number;
}

export interface CityPageProps extends Record<string, unknown> {
  city: string;
  listings: Listing[];
  // Add other properties as needed
}

export interface UnifiedListing {
  id: string;
  slug: string;
  title: string;
  city?: string;
  address_string?: string;
  eco_focus_tags?: string[];
  eco_notes_detailed?: string;
  // Add common properties from both Listing and SanityListing
}

// Add more types as needed
