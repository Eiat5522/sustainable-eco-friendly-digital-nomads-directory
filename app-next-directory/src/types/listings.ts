export interface PricingPlan {
  name: string;
  price: number;
  duration: string;
  features: string[];
}

export interface Coordinates {
  latitude: number | null;
  longitude: number | null;
}

export interface Listing {
  id: string;
  slug?: string;
  name: string;
  city: string;
  category?: 'coworking' | 'cafe' | 'accommodation';
  address_string: string;
  description_short: string;
  description_long: string;
  eco_focus_tags: string[];
  eco_notes_detailed: string;
  source_urls: string[];
  primary_image_url: string;
  gallery_image_urls: string[];
  digital_nomad_features: string[];
  last_verified_date: string;
  coordinates?: Coordinates;
  coworking_details?: {
    operating_hours: string | null;
    pricing_plans: PricingPlan[];
    specific_amenities_coworking: string[];
  };
  cafe_details?: {
    operating_hours: string;
    price_indication: string;
    menu_highlights_cafe: string[];
    wifi_reliability_notes: string;
  };
  accommodation_details?: {
    accommodation_type: string;
    price_per_night_thb_range: {
      min: number;
      max: number;
    };
    room_types_available: string[];
    specific_amenities_accommodation: string[];
  };
}
