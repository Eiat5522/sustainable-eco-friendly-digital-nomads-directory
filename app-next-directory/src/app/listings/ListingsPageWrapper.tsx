'use client';

import { useEffect, useState } from 'react';
import { type Listing } from '@/types/listings';
import { type SanityListing } from '@/types/sanity';
import ListingsPage from './ListingsPage';

export default function ListingsPageWrapper() {
  const [listings, setListings] = useState<Array<Listing | SanityListing>>([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    async function fetchListings() {
      try {
        const response = await fetch('/api/listings?featured=true');
        const data = await response.json();
        if (data.status === 'success') {
          setListings(data.data);
        }
      } catch (error) {
        console.error('Error fetching listings:', error);
      } finally {
        setIsLoading(false);
      }
    }

    fetchListings();
  }, []);

  if (isLoading) {
    return (
      <div className="min-h-screen bg-gray-50 flex items-center justify-center">
        <div className="text-center">
          <div className="w-16 h-16 border-4 border-primary-600 border-t-transparent rounded-full animate-spin mx-auto mb-4"></div>
          <p className="text-gray-600">Loading listings...</p>
        </div>
      </div>
    );
  }

  return <ListingsPage initialListings={listings} />;
}
