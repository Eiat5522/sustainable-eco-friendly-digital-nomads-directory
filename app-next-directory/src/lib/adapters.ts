// This file adapts between different types of listing data formats
// For example, it helps convert listings.json data to the format expected by components
// or converts between the Listing type (from listing.ts) and the Listing type (from listings.ts)

import { Listing as SanityListing } from '@/types/listing';
import { Listing as JsonListing } from '@/types/listings';

/**
 * Converts a JSON listing format to the Sanity CMS listing format
 * @param jsonListing The listing from listings.json
 * @returns A listing in the Sanity format
 */
export function jsonToSanityListing(jsonListing: JsonListing): SanityListing {
  return {
    _id: jsonListing.id,
    name: jsonListing.name,
    slug: jsonListing.name.toLowerCase().replace(/[^a-z0-9]+/g, '-'),
    description: jsonListing.description_long,
    type: jsonListing.category as any, // Type conversion between the two formats
    priceRange: 'moderate', // Default value
    price_indication: 'moderate', // Required field
    mainImage: {
      asset: {
        _ref: 'image-reference',
        url: jsonListing.primary_image_url
      }
    },
    city: {
      _id: 'city-id',
      name: jsonListing.city,
      slug: jsonListing.city.toLowerCase().replace(/[^a-z0-9]+/g, '-'),
      listingCount: 0,
      country: 'Thailand'
    },
    ecoTags: jsonListing.eco_focus_tags.map((tag, index) => ({
      _id: `tag-${index}`,
      name: tag.replace(/_/g, ' '),
      slug: tag,
      description: '',
      listingCount: 0
    })),
    ecoRating: calculateEcoRating(jsonListing),
    location: (jsonListing as any).coordinates
              ? {
                  lat: (jsonListing as any).coordinates.latitude || 0,
                  lng: (jsonListing as any).coordinates.longitude || 0,
                  coordinates: [
                    (jsonListing as any).coordinates.latitude || 0,
                    (jsonListing as any).coordinates.longitude || 0
                  ]
                }
              : {
                  lat: 0,
                  lng: 0,
                  coordinates: [0, 0]
                },
    address: jsonListing.address_string,
    rating: 4.5, // Default rating
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  };
}

/**
 * Calculate an eco rating score between 0-100 based on the listing data
 */
export function calculateEcoRating(listing: JsonListing): number {
  // Base score starting at 50
  let score = 50;
  
  // Add points for each eco tag (max 30 points)
  score += Math.min(listing.eco_focus_tags.length * 10, 30);
  
  // Add points if it has detailed eco notes
  if (listing.eco_notes_detailed && listing.eco_notes_detailed.length > 50) {
    score += 10;
  }
  
  // Add points for digital nomad features
  if (listing.digital_nomad_features && listing.digital_nomad_features.length > 0) {
    score += 5;
  }
  
  // Cap the score at 100
  return Math.min(score, 100);
}
