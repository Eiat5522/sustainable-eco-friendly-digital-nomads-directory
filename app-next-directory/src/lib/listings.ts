// src/lib/listings.ts
import listings from '../data/listings.json';
import { Listing } from '../types/listings';

export function getListingsByCity(city: string): Listing[] {
  return listings.filter(
    (listing) =>
      !!listing.city &&
      listing.city.toLowerCase() === city.trim().toLowerCase() &&
      (listing.category === 'coworking' ||
        listing.category === 'cafe' ||
        listing.category === 'accommodation')
  ) as Listing[];
}

type FilterOptions = {
  city?: string;
  category?: 'coworking' | 'cafe' | 'accommodation';
  hasEcoTags?: boolean;
  hasDnFeatures?: boolean;
};

export function filterListings(options: FilterOptions): Listing[] {
  return listings.filter((listing) => {
    if (options.city) {
      if (
        !listing.city ||
        listing.city.toLowerCase() !== options.city.trim().toLowerCase()
      ) {
        return false;
      }
    }
    if (options.category) {
      if (listing.category !== options.category) {
        return false;
      }
    }
    if (options.hasEcoTags) {
      if (!listing.eco_focus_tags || listing.eco_focus_tags.length === 0) {
        return false;
      }
    }
    if (options.hasDnFeatures) {
      if (
        !listing.digital_nomad_features ||
        listing.digital_nomad_features.length === 0
      ) {
        return false;
      }
    }
    // Ensure category is valid for Listing type
    return (
      listing.category === 'coworking' ||
      listing.category === 'cafe' ||
      listing.category === 'accommodation'
    );
  }) as Listing[];
}
