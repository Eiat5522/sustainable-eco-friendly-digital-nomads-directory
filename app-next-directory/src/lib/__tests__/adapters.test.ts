import { jsonToSanityListing } from '../adapters';
import { Listing as JsonListing } from '@/types/listings';
import { Listing as SanityListing } from '@/types/listing';

// Mock Date for deterministic createdAt/updatedAt
const MOCK_DATE = '2023-01-01T00:00:00.000Z';
beforeAll(() => {
  jest.spyOn(global, 'Date').mockImplementation(() => ({
    toISOString: () => MOCK_DATE,
  } as any));
});
afterAll(() => {
  (Date as any).mockRestore && (Date as any).mockRestore();
});

describe('jsonToSanityListing', () => {
  const baseJsonListing: JsonListing = {
    id: 'abc123',
    name: 'Eco Space',
    city: 'Bangkok',
    category: 'coworking',
    address_string: '123 Green Rd',
    description_short: 'A short desc',
    description_long: 'A long eco-friendly description for digital nomads.',
    eco_focus_tags: ['solar_power', 'recycling'],
    eco_notes_detailed: 'We use solar panels and recycle all waste.',
    source_urls: ['https://example.com'],
    primary_image_url: 'https://img.com/eco.jpg',
    gallery_image_urls: ['https://img.com/eco1.jpg', 'https://img.com/eco2.jpg'],
    digital_nomad_features: ['fast_wifi', 'community_events'],
    last_verified_date: '2023-01-01',
    coworking_details: {
      operating_hours: '9-18',
      pricing_plans: [],
      specific_amenities_coworking: ['wifi', 'printer'],
    },
    cafe_details: undefined,
    accommodation_details: undefined,
  };

  it('should convert a full JsonListing to SanityListing format', () => {
    const result = jsonToSanityListing(baseJsonListing);
    expect(result).toMatchSnapshot();
    expect(result._id).toBe(baseJsonListing.id);
    expect(result.name).toBe(baseJsonListing.name);
    expect(result.slug).toBe('eco-space');
    expect(result.city.name).toBe('Bangkok');
    expect(result.city.slug).toBe('bangkok');
    expect(result.ecoTags.length).toBe(2);
    expect(result.ecoTags[0]).toMatchObject({
      _id: 'tag-0',
      name: 'solar power',
      slug: 'solar_power',
    });
    expect(result.ecoRating).toBeGreaterThan(0);
    expect(result.createdAt).toBe(MOCK_DATE);
    expect(result.updatedAt).toBe(MOCK_DATE);
  });

  it('should handle empty eco_focus_tags', () => {
    const listing = { ...baseJsonListing, eco_focus_tags: [] };
    const result = jsonToSanityListing(listing);
    expect(result.ecoTags).toEqual([]);
    expect(result.ecoRating).toBe(55); // base score + digital_nomad_features
  });

  it('should handle missing coordinates gracefully', () => {
    const listing = { ...baseJsonListing, coordinates: undefined as any };
    // coordinates is not in the type, but test for undefined
    const result = jsonToSanityListing(listing as any);
    // expect coordinates to default to [0, 0]
    expect(result.location.coordinates).toEqual([0, 0]);
  });

  it('should default priceRange and rating', () => {
    const result = jsonToSanityListing(baseJsonListing);
    expect(result.priceRange).toBe('moderate');
    expect(result.rating).toBe(4.5);
  });

  it('should generate slug with only valid characters', () => {
    const listing = { ...baseJsonListing, name: 'Eco!@# Space 2023' };
    const result = jsonToSanityListing(listing);
    expect(result.slug).toBe('eco-space-2023');
  });
});

import { calculateEcoRating } from '../adapters';

describe('calculateEcoRating', () => {
  const baseJsonListing: JsonListing = {
    id: 'abc123',
    name: 'Eco Space',
    city: 'Bangkok',
    category: 'coworking',
    address_string: '123 Green Rd',
    description_short: 'A short desc',
    description_long: 'A long eco-friendly description for digital nomads.',
    eco_focus_tags: [],
    eco_notes_detailed: '',
    source_urls: [],
    primary_image_url: '',
    gallery_image_urls: [],
    digital_nomad_features: [],
    last_verified_date: '2023-01-01',
    coworking_details: undefined,
    cafe_details: undefined,
    accommodation_details: undefined,
  };

  it('returns base score 50 for minimal listing', () => {
    expect(calculateEcoRating(baseJsonListing)).toBe(50);
  });

  it('adds 10 points per eco tag, max 30', () => {
    const listing = { ...baseJsonListing, eco_focus_tags: ['a', 'b', 'c', 'd'] };
    expect(calculateEcoRating(listing)).toBe(80); // 4 tags = 40, but capped at 30
  });

  it('adds 10 points for detailed eco notes > 50 chars', () => {
    const listing = { ...baseJsonListing, eco_notes_detailed: 'x'.repeat(51) };
    expect(calculateEcoRating(listing)).toBe(60);
  });

  it('adds 5 points for digital nomad features', () => {
    const listing = { ...baseJsonListing, digital_nomad_features: ['wifi'] };
    expect(calculateEcoRating(listing)).toBe(55);
  });

  it('caps the score at 100', () => {
    const listing = {
      ...baseJsonListing,
      eco_focus_tags: ['a', 'b', 'c', 'd'],
      eco_notes_detailed: 'x'.repeat(51),
      digital_nomad_features: ['wifi'],
    };
    expect(calculateEcoRating(listing)).toBe(95);
    // Add more features to push over 100
    const overListing = {
      ...listing,
      eco_focus_tags: ['a', 'b', 'c', 'd', 'e', 'f'],
      digital_nomad_features: ['wifi', 'community', 'events'],
    };
    expect(calculateEcoRating(overListing)).toBe(95); // max possible score is 95 due to eco tag cap
  });
});
