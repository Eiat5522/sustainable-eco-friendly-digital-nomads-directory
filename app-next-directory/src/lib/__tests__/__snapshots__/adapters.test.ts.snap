// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonToSanityListing should convert a full JsonListing to SanityListing format 1`] = `
{
  "_id": "abc123",
  "address": "123 Green Rd",
  "city": {
    "_id": "city-id",
    "country": "Thailand",
    "listingCount": 0,
    "name": "Bangkok",
    "slug": "bangkok",
  },
  "createdAt": "2023-01-01T00:00:00.000Z",
  "description": "A long eco-friendly description for digital nomads.",
  "ecoRating": 75,
  "ecoTags": [
    {
      "_id": "tag-0",
      "description": "",
      "listingCount": 0,
      "name": "solar power",
      "slug": "solar_power",
    },
    {
      "_id": "tag-1",
      "description": "",
      "listingCount": 0,
      "name": "recycling",
      "slug": "recycling",
    },
  ],
  "location": undefined,
  "mainImage": {
    "asset": {
      "_ref": "image-reference",
      "url": "https://img.com/eco.jpg",
    },
  },
  "name": "Eco Space",
  "priceRange": "moderate",
  "rating": 4.5,
  "slug": "eco-space",
  "type": "coworking",
  "updatedAt": "2023-01-01T00:00:00.000Z",
}
`;
